quil-rs/src/quil.rs

```diff
@@ -0,0 +1,71 @@
+/// A trait to wrap items which represent some construct within the Quil language.
+///
+/// If you want to serialize an object to string and fail if it can't be represented as valid Quil, then use
+/// `to_quil()`. If you want to serialize an object to string infallibly, and can tolerate invalid Quil, then
+/// use `to_quil_or_debug()`.
+pub trait Quil {
+    /// Return a string in valid Quil syntax or an error if the item cannot be represented with valid Quil.
+    fn to_quil(&self) -> Result<String, ToQuilError> {
+        let mut buffer = String::new();
+        self.write(&mut buffer, false)?;
+        Ok(buffer)
+    }
+
+    /// Return a string in valid Quil syntax if possible. Any individual component of this object
+    /// which cannot be represented in Quil will be replaced with a `Debug` representation of that
+    /// component.
+    fn to_quil_or_debug(&self) -> String
+    where
+        Self: std::fmt::Debug,
```

Comment:
Having the `Debug` restriction on the method feels weird. I would prefer it on the trait itself rather than the method:

```suggestion
pub trait Quil: std::fmt::Debug {
    /// Return a string in valid Quil syntax or an error if the item cannot be represented with valid Quil.
    fn to_quil(&self) -> Result<String, ToQuilError> {
        let mut buffer = String::new();
        self.write(&mut buffer, false)?;
        Ok(buffer)
    }

    /// Return a string in valid Quil syntax if possible. Any individual component of this object
    /// which cannot be represented in Quil will be replaced with a `Debug` representation of that
    /// component.
    fn to_quil_or_debug(&self) -> String
```